flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBaseTest.java

```diff
@@ -0,0 +1,352 @@
+                                return builder.addContactPoint("127.0.0.1")
+                                        .withoutJMXReporting()
+                                        .withoutMetrics()
+                                        .build();
+                            }
+                        },
+                        Integer.MAX_VALUE,
+                        DEFAULT_MAX_CONCURRENT_REQUESTS_TIMEOUT) {
+                    @Override
+                    public ListenableFuture send(Object value) {
+                        return null;
+                    }
+                };
+        cassandraOutputFormatBase.configure(new Configuration());
+        cassandraOutputFormatBase.open(1, 1);
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testSuccessfulWrite() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            testCassandraOutputFormat.enqueueCompletableFuture(
+                    CompletableFuture.completedFuture(null));
+
+            final int originalPermits = testCassandraOutputFormat.getAvailablePermits();
+            assertTrue(originalPermits > 0);
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+
+            testCassandraOutputFormat.writeRecord("hello");
+
+            Assert.assertEquals(originalPermits, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testThrowErrorOnClose() throws Exception {
+        TestCassandraOutputFormat testCassandraOutputFormat = createTestCassandraOutputFormat();
+        testCassandraOutputFormat.open(1, 1);
+
+        Exception cause = new RuntimeException();
+        testCassandraOutputFormat.enqueueCompletableFuture(
+                FutureUtils.completedExceptionally(cause));
+        testCassandraOutputFormat.writeRecord("hello");
+        try {
+            testCassandraOutputFormat.close();
+            Assert.fail("Close should have thrown an exception.");
+        } catch (IOException e) {
+            ExceptionUtils.findThrowable(e, candidate -> candidate == cause).orElseThrow(() -> e);
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testThrowErrorOnWrite() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            Exception cause = new RuntimeException();
+            testCassandraOutputFormat.enqueueCompletableFuture(
+                    FutureUtils.completedExceptionally(cause));
+
+            testCassandraOutputFormat.writeRecord("hello");
+
+            try {
+                testCassandraOutputFormat.writeRecord("world");
+                // should fail because only one completableFuture was enqueued
```

Comment:
This seems misleading; it's failing because the previous write has failed.

Reply:
:+1: